src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -637,38 +632,39 @@ internal MemberAccessor MemberAccessorStrategy
             }
         }
 
-        /// <summary>
-        /// JsonTypeInfo resolver used for serialization.
-        /// </summary>
-        internal IJsonTypeInfoResolver? _typeInfoResolver;
+        internal bool IsInitializedForReflectionSerializer { get; private set; }
 
         /// <summary>
         /// Initializes the converters for the reflection-based serializer.
         /// </summary>
         [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
         [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
         [MemberNotNull(nameof(_typeInfoResolver))]
-        internal void EnsureInitializedForReflectionSerializer()
+        internal void InitializeForReflectionSerializer()
         {
-            if (s_defaultTypeInfoResolver != null && _typeInfoResolver != null && (_serializerContext == null || _serializerContext.FallbackResolver != null))
+            RootConverters();
+
+            if (_typeInfoResolver is JsonSerializerContext ctx)
             {
-                return;
+                // .NET 6 backward compatibility; use fallback to reflection serialization
+                // TODO 1: Consider removing this behaviour (needs to be filed as a breaking change).
+                // TODO 2: This approach _changes_ the reported JsonTypeInfoResolver:
+                //         evaluate if this is the appropriate way to implement fallback.
+                _typeInfoResolver = JsonTypeInfoResolver.Combine(ctx, DefaultJsonTypeInfoResolver.DefaultInstance);
```

Comment:
this will actually make internal `SerializerContext` property return null now which might potentially change behavior in some places where it's not desired. I think we should keep reporting original context